### ISSUE TITLE
Joins fail with empty or nil data

### ISSUE THREAD

**Issue by gnilrets** [v0dro/daru#171]
Found two cases where joins are failing

1 - The left dataframe is empty:

````ruby
df1 = Daru::DataFrame.new({ a: [] }, order: [:a])
df2 = Daru::DataFrame.new({ a: [1,2,3], b: [10,20,30] })

df1.join(df2, on: [:a], how: :left)

#=> ArgumentError: comparison of NilClass with Array failed
````

2 - The left or right dataframe contains nil values for the join keys:

````ruby
df1 = Daru::DataFrame.new({ a: [1,nil,2] }, order: [:a])
df2 = Daru::DataFrame.new({ a: [1,2,3], b: [10,20,30] })

df1.join(df2, on: [:a], how: :left)

#=> ArgumentError: comparison of NilClass with Array failed
````

**Comment by zverok** [comment]
I'll fix this today. It was me who rewrite the join last time :(

**Comment by gnilrets** [comment]
Oh, I thought it was me.  I should mention that this was pre-0.1.3 (I'm using abf306cade053e5700cb3c43fcf83b4ad1cfaa30 from April 22).  I haven't tested this against the latest master branch because I didn't think much had changed with joins.

**Comment by zverok** [comment]
OK, in current master only one of two fails:
```ruby
df1 = Daru::DataFrame.new({ a: [] }, order: [:a])
df2 = Daru::DataFrame.new({ a: [1,2,3], b: [10,20,30] })
df1.join(df2, on: [:a], how: :left)
# => #<Daru::DataFrame(0x2)>
#   a   b 

df1 = Daru::DataFrame.new({ a: [1,nil,2] }, order: [:a])
df2 = Daru::DataFrame.new({ a: [1,2,3], b: [10,20,30] })

df1.join(df2, on: [:a], how: :left)
# => ArgumentError: comparison of Array with Array failed
```

Will look at it.

I've rewritten the join logic when tried to satisfy strict Rubocop requirements.

**Comment by gmega** [comment]
Hi there @zverok ,

   this snippet:

```ruby
df1 = Daru::DataFrame.new({ a: [1,nil,2] }, order: [:a])
df2 = Daru::DataFrame.new({ a: [1,2,3], b: [10,20,30] })

df1.join(df2, on: [:a], how: :left)
```

still seems to fail for me on 0.2.0 with `ArgumentError: comparison of Array with Array failed
`. I believe I have the right version installed:

```sh
➜  ~ gem list | grep daru
daru (0.2.0)
```

Going through the [test case that got added](https://github.com/SciRuby/daru/pull/177/commits/6069f7e27b1442ddab87ccc102cc5c5d08b1d324#diff-9a856a5cbb8265463fa1bb64f02258e1) with the fix for #171, I see that the test case passes as it is, but fails if I modify it to use a numeric vector in `:name` instead of a vector of strings; i.e., this works:

```ruby
left = Daru::DataFrame.new({
  :id   => [1, 2, 3, 4],
  :name => ['Pirate', nil, 'Ninja', 'Spaghetti']
})

right = Daru::DataFrame.new({
  :id => [1, 2, 3, 4],
  :name => ['Rutabaga', 'Pirate', 'Darth Vader', 'Ninja']
})

left.join(right, how: :outer, on: [:name])
```

but this doesn't:

```ruby
left = Daru::DataFrame.new({
  :id   => [1, 2, 3, 4],
  :name => [1, nil, 3, 4]
})

right = Daru::DataFrame.new({
  :id => [1, 2, 3, 4],
  :name => [1, 2, 3, 4]
})

left.join(right, how: :outer, on: [:name])
```

**Comment by zverok** [comment]
Funny. Will take a look. Thanks for digging it out.
